cql3/statements/describe_statement.cc

```diff
@@ -160,14 +160,80 @@ description type(replica::database& db, const lw_shared_ptr<keyspace_metadata>&
+ * If a UDT `a` is used in any field in a UDT `b`,
+ * then `a` should be before `b`.
+ *
+ * There shouldn't be any cycles as we don't allow any
+ * circular dependencies in UDTs.
+ *
+ * Algorithm:
+ * - for each UDT, count in how many other UDTs it's used
+ * - push to result UDTs which counter is 0 (they aren't used in any other UDTs)
+ * - for each UDT `u` in current result vector:
+ *   - get UDTs `ref_udt`s used in `u`
+ *   - decrement counter of `ref_udt`s
+ *   - if counter of any `ref_udt` drops to 0, push that `ref_udt` to end of result vector
+ * - reverse order of result
+ */
+static future<std::vector<shared_ptr<const keyspace_element>>> get_sorted_udts(const lw_shared_ptr<keyspace_metadata>& ks) {
+    auto& udts_map = ks->user_types().get_all_types();
+    auto get_udt = [&udts_map] (const sstring& name) {
+        return udts_map.at(to_bytes(name));
+    };
+    
+    // UDT's name -> in how many other UDTs is it used
+    std::map<user_type, size_t> usage_cnt_map;
+    // Contains names of UDTs sorted in a reversed order to final result.
+    // (each UDT will be pushed before its dependencies)
+    std::vector<user_type> sorted_udts;
+    sorted_udts.reserve(udts_map.size());
+
+    // Initialize usage counter map to 0 for all UDTs
+    for (auto& [_, udt]: udts_map) {
+        usage_cnt_map.insert({udt, 0});
+    }
+    // For each UDT, calculate how many other UDTs depend on it in their definitions.
+    for (auto& [_, udt]: udts_map) {
+        for (auto& ref_udt: udt->get_all_referenced_user_types()) {
+            usage_cnt_map[get_udt(ref_udt)]++;
+        }
+        co_await coroutine::maybe_yield();
+    }
+
+    // Firstly, push UDTs which are not dependencies in others
+    for (auto& [udt, cnt]: usage_cnt_map) {
+        if (cnt == 0) {
+            sorted_udts.push_back(udt);
+        }
+    }
+    
+    // When a UDT is in `sorted_udts`, we can "remove" it from the graph and decrease counter
+    // in all UDTs it's referencing.
+    // If an item's counter drops down to 0, it means all dependencies of that UDT 
+    // were already pushed to result vector, so we can add it to the end of the vector.
+    for (size_t i = 0; i < sorted_udts.size(); i++) {
+        for (auto& ref_udt_name: sorted_udts[i]->get_all_referenced_user_types()) {
+            auto ref_udt = get_udt(ref_udt_name);
+            if (--usage_cnt_map[ref_udt] == 0) {
+                sorted_udts.push_back(ref_udt);
+            }
+        }
+        co_await coroutine::maybe_yield();
+    }
+
+    // Cycles in UDTs graph are illegal, so it shouldn't happen
+    if (sorted_udts.size() != udts_map.size()) {
+        on_internal_error(dlogger, "cycles within UDTs detected");
+    }
 
-    co_return co_await generate_descriptions(db, user_types, (with_stmt) ? std::optional(true) : std::nullopt);
+    co_return boost::copy_range<std::vector<shared_ptr<const keyspace_element>>>(sorted_udts | boost::adaptors::reversed);
```

Comment:
It would make some sense to sort independent types by name (so we have a stable order). But not worth the effort now.